### PR TITLE
Make sure that base folder exists

### PR DESCRIPTION
**Pull request by airxnoor**
Crashlytics mentions a few crashes related to the base folder not ready for storing log files. This pr ensures that the base log folder is always present before a file operation starts.

### CODE REVIEW

File: src/commonMain/kotlin/com/airthings/lib/logging/facility/JsonLoggerFacility.kt [hunk 1]
```diff
@@ -221,14 +217,28 @@ class JsonLoggerFacility(
...
-    suspend fun files(date: LogDate): Collection<String> = io.of(baseFolder, date)
+    suspend fun files(date: LogDate): Collection<String> {
+        ensureBaseFolder()
+        return io.of(baseFolder, date)
+    }
+
+    private suspend fun ensureBaseFolder() {
+        // Please note: The call to `io.mkdirs()` returns true if the directory exists, which may be
+        // different from the platform's implementation.
+        if (!io.mkdirs(baseFolder)) {
```
Comment (jstarczewski) [inline src/commonMain/kotlin/com/airthings/lib/logging/facility/JsonLoggerFacility.kt]: This PR does not remove the race condition created by usage of a coroutine in `init` block as that's the of the bug basically 

Also we have to remember to always be sure the directory was created/we tried to create the directory which I think we should not relay on developer so heavily. I suggest an approach which will check whether the directory is created before performing any `io` operations for example introducing the following decorator design pattern

```Kotlin
internal class PlatformInDirFileInputOutput(
    private val dirPath: String,
    private val platformFileInputOutput: PlatformFileInputOutput = PlatformFileInputOutputImpl(),
) : PlatformFileInputOutput {

    override val pathSeparator: Char = platformFileInputOutput.pathSeparator

    override suspend fun mkdirs(path: String): Boolean = ensureDir { mkdirs(path) }

    override suspend fun size(path: String): Long = ensureDir { size(path) }

    override suspend fun write(path: String, position: Long, contents: String) =
        ensureDir { write(path, position, contents) }

    override suspend fun append(path: String, contents: String) = ensureDir { append(path, contents) }

    override suspend fun ensure(path: String) = ensureDir { ensure(path) }

    override suspend fun delete(path: String) = ensureDir { delete(path) }

    override suspend fun of(path: String): Collection<String> = ensureDir { of(path) }

    override suspend fun of(path: String, date: LogDate): Collection<String> = ensureDir {
        of(path, date)
    }

    private suspend fun <T> ensureDir(
        action: suspend PlatformFileInputOutput.() -> T,
    ): T = with(platformFileInputOutput) {
        mkdirs(dirPath)
        platformFileInputOutput.action()
    }
}
```

Then we can safely remove 
1. `init` block from `FileLoggerFacility` 
2. Use the new class like this `    private val io: PlatformFileInputOutput = PlatformInDirFileInputOutput(baseFolder)
`
3. Remove `ensureBaseFolder` function and it's usage where needed because all of the I/O operations are now safe/are possibly safe

4. Remove crashing from the library via explicit `throw` calls (this is debatable, as I think the lib should not crash the whole application, maybe we can introduce a lambda that will be invoked on such operations, for example one reporting to crashlytics etc.)  

So basically now we remove the race condition and we try to always be sure we have the needed folder when we do `io` operations

I will be testing my proposal to see if it works properly, I think it is more idiomatic to use as we are safe to do IO operations on a file if we **have to be sure it is in the proper directory** or we can use the old `PlatformFileInputOutputImpl` which is more generic, moreover it is easier to introduce in in every logger facility implementation as we have to swap the impl of the interface that is used in them (swap the `private val platformFileInputOutput: PlatformFileInputOutput = PlatformFileInputOutputImpl()`)

Comment (jstarczewski) [reply src/commonMain/kotlin/com/airthings/lib/logging/facility/JsonLoggerFacility.kt]: Ofc other enhancements can also be done

**Review by jstarczewski — COMMENTED** [review]
I lurked into implementation a bit more, please let me know what you think!

File: src/commonMain/kotlin/com/airthings/lib/logging/facility/JsonLoggerFacility.kt (hunk 1 above)
Comment (airxnoor) [reply src/commonMain/kotlin/com/airthings/lib/logging/facility/JsonLoggerFacility.kt]: Agreed, this looks much better and serves the purpose! I'll refactor according to your code. Thanks for sharing and helping out 💪

Comment (jstarczewski) [reply src/commonMain/kotlin/com/airthings/lib/logging/facility/JsonLoggerFacility.kt]: Great, thank you!

Comment (airxnoor) [reply src/commonMain/kotlin/com/airthings/lib/logging/facility/JsonLoggerFacility.kt]: @jstarczewski I've refactored the code, please take another look.

File: src/commonMain/kotlin/com/airthings/lib/logging/facility/FileLoggerFacility.kt [hunk 2]
```diff
@@ -128,19 +129,15 @@ class FileLoggerFacility(
         notifier = null,
     )
 
-    private val io: PlatformFileInputOutput = PlatformFileInputOutputImpl()
+    private val io: PlatformFileInputOutput = DelegateFileInputOutput(
+        folder = baseFolder,
+        io = PlatformFileInputOutputImpl(),
+        onFolderMissing = {
+            notifier?.onLogFolderInvalid(it) ?: DelegateFileInputOutput.reportMissingFolder(it)
```
Comment (airxnoor) [inline src/commonMain/kotlin/com/airthings/lib/logging/facility/FileLoggerFacility.kt:136]: @jstarczewski I kept the previous logic where an exception is thrown, for backward-compatibility.

**Review by jstarczewski — APPROVED** [review]
Great job! If there are no changes needed for other logger facilities then let's get that merge and please provide a PR to update the KMP log on app side if you can, or let me know so I will do it  :)
